### PR TITLE
redis: generalize connection pool and routing

### PR DESCRIPTION
**Pull request by maximebedard**
Description: In order to build a [memcached filter](https://github.com/envoyproxy/envoy/issues/5997#issuecomment-488060827) with similar functionality to the redis proxy, I thought it would be a good idea to generalize the connection pool and the router. This PR only moved theses primitives into a new module `Common::Multiplexing`. A follow-up PR will address removing the dependency to the redis-codec.
Risk Level: low
Testing: Moved unit tests
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
cc @derekargueta for first pass. Are you two talking since I think @derekargueta is working on this also?

cc @HenryYYang and @msukalski also for a look.

**Comment by maximebedard** [comment]
I have chatted with @derekargueta, but we have slightly different requirements (ASCII codec vs Binary codec) for memcached, but I think we can share a lot of the primitives currently being used by redis before we even start to write the memcached filter. I've experimented a bit [here](https://github.com/envoyproxy/envoy/commit/3dc28c09a4b20a4ff95f2e3fce6670d7978a4b58) to see how this generalization would look like, any design feedback would be truly appreciated!

**Comment by derekargueta** [comment]
Change LGTM too, thanks @maximebedard.

Re sharing ConnPool and Router, the ideal end goal I think we're going for is something twemproxy-like in terms of shared connection infrastructure, where we have an abstracted connection/routing layer that can just pipe through any protocol (particularly for kv-stores). We'll definitely have to introduce some intermediate types and factories but I think that'll be worth it rather than re-implementing everything again.

@maximebedard let's sync up offline on next steps

File: test/extensions/filters/network/common/multiplexing/BUILD [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+envoy_cc_mock(
+    name = "multiplexing_mocks",
+    srcs = ["mocks.cc"],
+    hdrs = ["mocks.h"],
+    deps = [
+        "//source/common/common:assert_lib",
+        "//source/extensions/filters/network/common/multiplexing:conn_pool_interface",
+        "//source/extensions/filters/network/common/multiplexing:router_interface",
+        "//source/extensions/filters/network/common/redis:client_interface",
```
Comment (mattklein123) [inline test/extensions/filters/network/common/multiplexing/BUILD]: It seems wrong that we are still pointing into Redis in here? Can we tease that apart?

Comment (maximebedard) [reply test/extensions/filters/network/common/multiplexing/BUILD]: I was thinking of removing that dependency in a commit similar to this one https://github.com/envoyproxy/envoy/commit/3dc28c09a4b20a4ff95f2e3fce6670d7978a4b58. I can merge it into this branch and fix the tests if it's preferred to do it all in one go. WDYT?

Comment (mattklein123) [reply test/extensions/filters/network/common/multiplexing/BUILD]: I would discuss with the redis folks in Slack and see what the consensus is. I'm fine with whatever you all want to do, but we should try to keep things clean which would mean no circular dependencies like this.

**Comment by maximebedard** [comment]
Closing. As we discussed on slack, the odds of the implementation details changing from a memcached filter vs a redis filter are pretty big and this is likely not worth generalizing now.

**Pull request closed without merging**